### PR TITLE
Port ting search carousel to SAL [BBS-178]

### PR DESCRIPTION
**Pull request by tripox**

### CODE REVIEW

File: modules/opensearch/includes/opensearch.search.inc [hunk 1]
```diff
@@ -109,19 +109,29 @@ function opensearch_search_search(TingSearchRequest $ting_query) {
...
-
   // Add field filter.
   if (!empty($ting_query->getFieldFilters())) {
     $render = new StatementGroupRender(new DingProviderStrategy());
     $field_filters = $render->renderStatements($ting_query->getFieldFilters());
     $query_parts[] = $field_filters;
   }
 
+  // Add material filter, this has to be the very last thing we do as we have
+  // to handle the filter differently query_parts is empty.
```
Comment (kasperg) [inline modules/opensearch/includes/opensearch.search.inc]: I really appreciate the comment but I do not see how it is implemented. I would have expected a check whether `$query_parts` is `empty()` or not.

Either the comment is no longer relevant and should be removed or the subsequent code needs to be updated.

File: modules/ting_search_carousel/ting_search_carousel.module [hunk 2]
```diff
@@ -283,31 +283,30 @@ function _ting_search_carousel_find_entities_with_covers($query, $start, $size,
...
+    ->setPage($start)
+    ->setCount($size)
+    ->setPopulateCollections(FALSE)
+    ->setMaterialFilter($ignore, FALSE);
 
-  // TODO: use the Search Abstraction Layer (ting_start_query()).
-  // Ting pages are 1, not 0, based.
-  $result = opensearch_do_search($query, ($start++) + 1, $size, $options);
+  $results = $sal_query->execute();
+  $collections = $results->getTingEntityCollections();
```
Comment (kasperg) [inline modules/ting_search_carousel/ting_search_carousel.module]: I would expect `$results->getTingEntityCollections()` to be moved to `L305`.

The check on `L297` does not make any sense if we already assume on `L294` that the return value from `$sal_query->execute()` is a proper object.

File: modules/opensearch/includes/opensearch.search.inc [hunk 3]
```diff
@@ -109,19 +109,29 @@ function opensearch_search_search(TingSearchRequest $ting_query) {
...
+  if (!empty($material_filter)) {
+    if ($ting_query->isMaterialFilterInclude()) {
+      // The material filter is a list of IDs to include. Add OR conditions.
+      $query_parts[] = implode(' OR ', $material_filter);
+    }
+    else {
+      // We're excluding, meaning we need a NOT in front of all id's. We can't
+      // start the query with a NOT so we add a wildcard match to the start of
+      // the filter if query is empty.
+      $query_parts[] = '* NOT ' . implode(' NOT ', $material_filter);
```
Comment (kasperg) [inline modules/opensearch/includes/opensearch.search.inc]: Nice level of commenting here.

File: modules/ding_provider/tests/ding_provider_implementation.test [hunk 4]
```diff
@@ -569,6 +569,7 @@ abstract class DingSearchProviderImplementationTestCase extends DingProviderImpl
     $query = ding_provider_invoke('search', 'prepare_reference_query', $test_query_3);
     $this->assertFalse($query->isFuzzy());
     $this->assertEqual($query->getMaterialFilter(), [$test_query_3]);
+    $this->assertTrue($query->isMaterialFilterInclude());
```
Comment (kasperg) [inline modules/ding_provider/tests/ding_provider_implementation.test]: Would it be possible to extend the test with a case where the material filter is exclude?

**Review by kasperg — CHANGES_REQUESTED** [review]
Reviewed. I have bunch of questions and comments.

**Comment by tripox** [comment]
Done.
@kasperg added test but was not able to run it.
Running tests takes forever on my machine.

**Comment by kasperg** [comment]
@tripox: Tests fail 😢 

`Fatal error: Call to undefined method Connie\Search\ConnieSearchResult::setMaterialFilter() in /tmp/drupal/profiles/ding2/modules/ding_provider/tests/ding_provider_implementation.test on line 598`

**Review by danquah — APPROVED** [review]

**Pull request merged**